src/engine/channelmixer.cpp

```diff
@@ -0,0 +1,91 @@
+        CSAMPLE* pOutput,
+        const ChannelHandle& outputHandle,
+        unsigned int iBufferSize,
+        unsigned int iSampleRate,
+        EngineEffectsManager* pEngineEffectsManager) {
+    // Signal flow overview:
+    // 1. Clear pOutput buffer
+    // 2. Calculate gains for each channel
+    // 3. Pass each channel's calculated gain and input buffer to pEngineEffectsManager, which then:
+    //     A) Copies each channel input buffer to a temporary buffer
+    //     B) Applies gain to the temporary buffer
+    //     C) Processes effects on the temporary buffer
+    //     D) Mixes the temporary buffer into pOutput
+    // The original channel input buffers are not modified.
+    SampleUtil::clear(pOutput, iBufferSize);
+    ScopedTimer t("EngineMaster::applyEffectsAndMixChannels");
+    for (int i = 0; i < activeChannels->size(); ++i) {
+        EngineMaster::ChannelInfo* pChannelInfo = activeChannels->at(i);
+        EngineMaster::GainCache& gainCache = (*channelGainCache)[pChannelInfo->m_index];
+        CSAMPLE_GAIN oldGain = gainCache.m_gain;
+        CSAMPLE_GAIN newGain;
+        if (gainCache.m_fadeout) {
+            newGain = 0;
+            gainCache.m_fadeout = false;
+        } else {
+            newGain = gainCalculator.getGain(pChannelInfo);
+        }
+        gainCache.m_gain = newGain;
+        pEngineEffectsManager->processPostFaderAndMix(pChannelInfo->m_handle,
+                outputHandle,
+                pChannelInfo->m_pBuffer,
+                pOutput,
+                iBufferSize,
+                iSampleRate,
+                pChannelInfo->m_features,
+                oldGain,
+                newGain);
+    }
+}
+
+void ChannelMixer::applyEffectsInPlaceAndMixChannels(
+        const EngineMaster::GainCalculator& gainCalculator,
+        QVarLengthArray<EngineMaster::ChannelInfo*, kPreallocatedChannels>*
+                activeChannels,
+        QVarLengthArray<EngineMaster::GainCache, kPreallocatedChannels>*
+                channelGainCache,
+        CSAMPLE* pOutput,
+        const ChannelHandle& outputHandle,
+        unsigned int iBufferSize,
+        unsigned int iSampleRate,
+        EngineEffectsManager* pEngineEffectsManager) {
+    // Signal flow overview:
+    // 1. Calculate gains for each channel
+    // 2. Pass each channel's calculated gain and input buffer to pEngineEffectsManager, which then:
+    //    A) Applies the calculated gain to the channel buffer, modifying the original input buffer
+    //    B) Applies effects to the buffer, modifying the original input buffer
+    // 4. Mix the channel buffers together to make pOutput, overwriting the pOutput buffer from the last engine callback
+    ScopedTimer t("EngineMaster::applyEffectsInPlaceAndMixChannels");
+    SampleUtil::clear(pOutput, iBufferSize);
+    for (int i = 0; i < activeChannels->size(); ++i) {
+        EngineMaster::ChannelInfo* pChannelInfo = activeChannels->at(i);
```

Comment:
consider replacing with ranged-for, might result in some performance benefit even.

Reply:
How could that provide a performance benefit?

Reply:
Compilers may be able to optimize better when conveying the intent better.

Reply:
For example, currently `activeChannels->size()` is probably called on every iteration while the ranged-for implementation would only compare the begin- and end-pointers of the iterator. I'm not saying this is a huge benefit, I just believe that we should do what compiler writers expect from us so we get better code and they can possibly produce better binaries.

Reply:
Readability would benefit form a range based for loop. That's the main advantage, everything else is just bonus.

Reply:
https://github.com/Be-ing/mixxx/pull/59